[BLOG] doctesting for pydata libraries (#780)
Quansight/Quansight-website: apps/labs/posts/doctesting-for-pydata-libraries.md

---
title: 'Doctesting for PyData Libraries'

> author: [Sheila-nk]

Reviewer: ```suggestion
authors: [sheila-kahwai]
```

Added you to Storyblok, this is the generated slug. :)

published: September 30, 2023
description: 'The journey of a PyData Newbie'

> category: [Doctests]

Reviewer: ```suggestion
category: [Developer workflows]
```

We haven't defined a `Doctests` category for the blog, so suggesting `Developer workflows` - which I think is a good fit here.

    ```python
    >>> from scipy import linalg
    >>> from scpdt import testmod
    >>> res, hist = testmod(linalg, strategy='api')
    >>> res
    TestResults(failed=0, attempted=764)
    ```

But here's the twist: Scpdt could only perform doctesting on SciPy's and NumPy's public modules through a helper script, and that wasn't ideal. So, guess who stepped in to bridge the gap?

## Bridging the Gap with Pytest
[Pytest](https://docs.pytest.org/en/stable/index.html) already has a [doctesting module](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/doctest.py), but unfortunately, it doesn't meet the specific needs of the PyData libraries. Therefore, the crucial task was to ensure pytest could leverage the power of Scpdt for doctesting. This involved overriding some of doctest's functions and classes to incorporate scpdt's alternative doctesting objects. It also meant modifying pytest's behavior by implementing hooks, primarily for initialization and collection.

> Once all the technical juggling was done, it was time for what my mentor called "[dogfooding](https://www.forbes.com/sites/michaeldefranco/2014/03/04/not-eating-your-own-dog-food-you-probably-should-be-2/?sh=69c002d0692e)"(a term he picked up from Joel Spolsky's [essay](https://www.joelonsoftware.com/2001/05/05/what-is-the-work-of-dogs-in-this-country/)). The term simply means putting your own product to the test by using it, and I had to make sure that the plugin functioned as expected. I did this by locally running doctests on SciPy's modules. It was an eye-opener, exposing issues like faulty collection – for example, the plugin wasn't collecting compiled and NumPy universal functions for doctesting.

Reviewer: ```suggestion
Once all the technical juggling was done, it was time for what my mentor called "[dogfooding](https://www.forbes.com/sites/michaeldefranco/2014/03/04/not-eating-your-own-dog-food-you-probably-should-be-2/?sh=69c002d0692e)" (a term he picked up from Joel Spolsky's [essay](https://www.joelonsoftware.com/2001/05/05/what-is-the-work-of-dogs-in-this-country/)). The term simply means putting your own product to the test by using it, and I had to make sure that the plugin functioned as expected. I did this by locally running doctests on SciPy's modules. It was an eye-opener, exposing issues like faulty collection – for example, the plugin wasn't collecting compiled and NumPy universal functions for doctesting.
```

With the bugs and vulnerabilities exposed during this process, I was able to refine the plugin further. I then created a [pull request](https://github.com/scipy/scipy/pull/19242) to demonstrate how the pytest plugin could be seamlessly integrated into SciPy. The process is fairly straightforward:

1. **Installation**: Install the plugin via pip.
2. **Configuration**: Customize your doctesting through a `conftest.py` file.
3. **Running Doctests in SciPy**: If you're running doctests on SciPy, execute the command `python dev.py test --doctests` in your shell.
4. **Running Doctests on Other Packages**: If you're not working with SciPy, use the command `pytest --pyargs <your-package> --doctest-modules` to run your doctests.

> Voila! :tada: 

Reviewer: ```suggestion
Voila! 🎉
```

<p align="center">
    <img
     alt="An image featuring Kamala Harris on a phone call, with the phrase 'We did it, Joe' displayed at the bottom of the image. Adjacent to the image are the pytest logo, a plus sign, and the text representing the doctesting package 'SCPDT'."
     src="/posts/doctesting-for-pydata-libraries/we-did-it.png" />
</p>

## Future Goals

> I am currently in the process of integrating the plugin into SciPy; for more details, you can check out the PR linked [here](https://github.com/scipy/scipy/pull/19242). Looking ahead, our goal is to publish the plugin on [PyPI](https://pypi.org/) and extend its integration to NumPy and other PyData libraries. 

Reviewer: ```suggestion
I am currently in the process of integrating the plugin into SciPy; for more details, you can check out [the PR](https://github.com/scipy/scipy/pull/19242). Looking ahead, our goal is to publish the plugin on [PyPI](https://pypi.org/) and extend its integration to NumPy and other PyData libraries. 
```

For better accessibility, we try to ensure the link text has context for the destination. :)

If you run into challenges with floating-point arithmetic, face output issues related to whitespace and array abrreviations, need to validate example source code without output testing, or simply desire a customized doctesting experience, consider giving this [plugin](https://github.com/ev-br/scpdt) a try.

## The Journey's End
Throughout this incredible journey, I cherished every moment spent working, learning from my mentors: [Evgeni Burovski](https://github.com/ev-br) and [Melissa Weber Mendonça](https://github.com/melissawm), and being part of the Quansight team. I'm incredibly grateful for this opportunity, and I look forward to continuing my contributions to the pytest plugin even after the internship.

> Curious? Check out the plugin [here](https://github.com/ev-br/scpdt). Feel free to contribute – the more, the merrier! 🚀🐍

Reviewer: ```suggestion
Curious? Check out the [plugin repository on GitHub](https://github.com/ev-br/scpdt). Feel free to contribute – the more, the merrier! 🚀🐍
```
